pkg/webhook/node/plugins/cpunormalization/node_cpu_normalization.go

```diff
@@ -0,0 +1,120 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpunormalization
+
+import (
+	"context"
+	"fmt"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+	"github.com/koordinator-sh/koordinator/pkg/webhook/node/plugins"
+)
+
+var (
+	CPUNormalizationResourceList = []corev1.ResourceName{
+		corev1.ResourceCPU,
+		extension.BatchCPU,
+		extension.MidCPU,
+	}
+)
+
+func NewPlugin() *NodeCPUNormalizationPlugin {
+	plugin := &NodeCPUNormalizationPlugin{}
+	return plugin
+}
+
+var _ plugins.NodePlugin = &NodeCPUNormalizationPlugin{}
+
+type NodeCPUNormalizationPlugin struct {
+}
+
+func (n *NodeCPUNormalizationPlugin) Name() string {
+	return "NodeCPUNormalizationPlugin"
+}
+
+func (n *NodeCPUNormalizationPlugin) Validate(ctx context.Context, req admission.Request, node, oldNode *corev1.Node) error {
+	return fmt.Errorf("not implemented")
+}
+
+// Admit makes an admission decision based on the request attributes
+func (n *NodeCPUNormalizationPlugin) Admit(ctx context.Context, req admission.Request, node, oldNode *corev1.Node) error {
+	klog.V(3).Infof("enter NodeNormalizationPlugin plugin admit for %s", req.AdmissionRequest.Name)
+
+	op := req.AdmissionRequest.Operation
+	switch op {
+	case admissionv1.Create:
+		return nil
+	case admissionv1.Update:
+		return n.handleUpdate(node)
+	}
+	return nil
+}
+
+func (n *NodeCPUNormalizationPlugin) handleUpdate(node *corev1.Node) error {
+	if node.Annotations == nil || node.Annotations[extension.AnnotationCPUNormalizationRatio] == "" || node.Status.Allocatable == nil {
+		// do nothing
+		return nil
+	}
+
+	// parse cpu normalization ratio
+	cpuNormalizationRatio, err := extension.GetCPUNormalizationRatio(node)
+	if err != nil {
+		klog.Errorf("NodeNormalizationPlugin plugin admit for %s, failed to parse normalization: %v", node.Name, err)
+		return nil
+	}
+
+	if !extension.IsCPUNormalizationRatioDifferent(1.0, cpuNormalizationRatio) {
+		// remove the original allocatable
+		delete(node.Annotations, extension.AnnotationOriginalNodeAllocable)
+		// do nothing, and kubelet will update the allocatable resource.
+		return nil
+	}
+
+	// save original allocatable
+	originalAllocatable := corev1.ResourceList{}
+	for _, resourceName := range CPUNormalizationResourceList {
+		resourceValue, ok := node.Status.Allocatable[resourceName]
+		if ok {
+			originalAllocatable[resourceName] = resourceValue
+		}
+	}
+	if node.Annotations == nil {
+		node.Annotations = map[string]string{}
+	}
+	err = extension.SetOriginalNodeAllocatable(node.Annotations, originalAllocatable)
```

Comment:
The status mutating webhook cannot mutate fields other than status IIRC.

Reply:
It should be possible, we can verify it

Reply:
After some researching, I found that status subresource update to Node **can** update metadata, so it works.